pkg/printer/printer.go

```diff
@@ -384,6 +384,139 @@ func (p *Printer) WriteProtoNodeStatusEvent(r *observerpb.GetFlowsResponse) erro
+		if a := e.GetAgentStart(); a != nil {
+			return fmt.Sprintf("start time: %s", fmtTimestamp(a.Time))
+		}
+	case pb.AgentEventType_POLICY_UPDATED, pb.AgentEventType_POLICY_DELETED:
+		if p := e.GetPolicyUpdate(); p != nil {
+			return fmt.Sprintf("labels: [%s], revision: %d, count: %d",
+				strings.Join(p.Labels, ","), p.Revision, p.RuleCount)
+		}
+	case pb.AgentEventType_ENDPOINT_REGENERATE_SUCCESS, pb.AgentEventType_ENDPOINT_REGENERATE_FAILURE:
+		if r := e.GetEndpointRegenerate(); r != nil {
+			str := fmt.Sprintf("id: %d, labels: [%s]", r.Id, strings.Join(r.Labels, ","))
+			if re := r.Error; re != "" {
+				str += fmt.Sprintf(", error: %s", re)
+			}
+			return str
+		}
+	case pb.AgentEventType_ENDPOINT_CREATED, pb.AgentEventType_ENDPOINT_DELETED:
+		if ep := e.GetEndpointUpdate(); ep != nil {
+			str := fmt.Sprintf("id: %d", ep.Id)
+			if n := ep.PodName; n != "" {
+				str += fmt.Sprintf(", pod name: %s", n)
+			}
+			if n := ep.Namespace; n != "" {
+				str += fmt.Sprintf(", namespace: %s", n)
+			}
+			return str
+		}
+	case pb.AgentEventType_IPCACHE_UPSERTED, pb.AgentEventType_IPCACHE_DELETED:
+		if i := e.GetIpcacheUpdate(); i != nil {
+			str := fmt.Sprintf("cidr: %s, identity: %d", i.Cidr, i.Identity)
+			if i.OldIdentity != nil {
+				str += fmt.Sprintf(", old identity: %d", i.OldIdentity.Value)
+			}
+			if i.HostIp != "" {
+				str += fmt.Sprintf(", host ip: %s", i.HostIp)
+			}
+			if i.OldHostIp != "" {
+				str += fmt.Sprintf(", old host ip: %s", i.OldHostIp)
+			}
+			return str + fmt.Sprintf(", encrypt key: %d", i.EncryptKey)
+		}
+	case pb.AgentEventType_SERVICE_UPSERTED:
+		if svc := e.GetServiceUpsert(); svc != nil {
+			str := fmt.Sprintf("id: %d", svc.Id)
+			if fe := svc.FrontendAddress; fe != nil {
+				str += fmt.Sprintf(", frontend: %s", formatServiceAddr(fe))
+			}
+			if bes := svc.BackendAddresses; len(bes) != 0 {
+				backends := make([]string, 0, len(bes))
+				for _, a := range bes {
+					backends = append(backends, formatServiceAddr(a))
+				}
+				str += fmt.Sprintf(", backends: [%s]", strings.Join(backends, ","))
+			}
+			if t := svc.Type; t != "" {
+				str += fmt.Sprintf(", type: %s", t)
+			}
+			if tp := svc.TrafficPolicy; tp != "" {
+				str += fmt.Sprintf(", traffic policy: %s", tp)
+			}
+			if n := svc.Name; n != "" {
+				str += fmt.Sprintf(", name: %s", n)
+			}
+			if ns := svc.Namespace; ns != "" {
+				str += fmt.Sprintf(", namespace: %s", ns)
+			}
+			return str
+		}
+	case pb.AgentEventType_SERVICE_DELETED:
+		if s := e.GetServiceDelete(); s != nil {
+			return fmt.Sprintf("id: %d", s.Id)
+		}
+	}
+	return "UNKNOWN"
+}
+
+// WriteProtoAgentEvent writes v1.AgentEvent into the output writer.
+func (p *Printer) WriteProtoAgentEvent(r *observerpb.GetFlowsResponse) error {
+	e := r.GetAgentEvent()
+	if e == nil {
+		return errors.New("not an agent event")
+	}
+
+	switch p.opts.output {
+	case JSONOutput, JSONPBOutput:
+		return json.NewEncoder(p.opts.werr).Encode(r)
```

Comment:
Any reason to write to the error output here (i.e use `p.opts.werr` over `p.opts.w`)? It made my `hubble observe -t agent | jq` quite sad.

Reply:
I guess I must have copy&pasted this from `WriteProtoNodeStatusEvent` above without thinking much. In fact I think we could just use `return p.jsonEncoder.Encode(r)` here. Which makes me realize that we probably want to distinguish `JSONOuutput` and `JSONPBOutput`. How about the following?

```go
        switch p.opts.output {
        case JSONOutput:
                return p.jsonEncoder.Encode(e)
        case JSONPBOutput:
                return p.jsonEncoder.Encode(r)
        // ...
```

where `e` is the `observerpb.AgentEvent` and `r` is the `observerpb.GetFlowsResponse`.